### PR TITLE
New client implementation

### PR DESCRIPTION
**Pull request by jgraef**
## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I have resolved any merge conflicts.

## What's in this pull request?

Integrated RPC and metrics server into new implementation `nimiq-client2`. Also removed the old implementations and renamed `nimiq-{client, lib}2` to `nimiq-{client, lib}`.

### CODE REVIEW

**Comment by jgraef** [comment]
Warnings and Websocket-RPC related stuff will be fixed in next PR. But I can also put them into this PR.

File: client/nimiq-client.service [hunk 1]
```diff
@@ -1,15 +0,0 @@
-[Unit]
```
Comment (paberr) [inline client/nimiq-client.service:1]: The service file is actually quite handy, but you already mentioned that you'll reintroduce it in another PR.

File: lib/src/config/config.rs [hunk 2]
```diff
@@ -494,6 +626,25 @@ impl ClientConfigBuilder {
...
+    }
+
+    /// Adds a custom seed url
+    pub fn seed_list(&mut self, url: Url, public_key_opt: Option<PublicKey>) -> &mut Self {
+        self.seed(Seed::new_list(SeedList::new(url, public_key_opt)))
+    }
+
+    /// Sets the validator config. Since there is no configuration for validators (except key file)
+    /// yet, this will just enable the validator.
     pub fn validator(&mut self) -> &mut Self {
```
Comment (paberr) [inline lib/src/config/config.rs:648]: This probably needs to go behind a feature flag as the `validator` field itself is behind a feature flag, too.

The `Some(Some(ValidatorConfig{}))` part looks strange, but I guess the extra Option comes from the builder.
`ValidatorConfig{}` can be simplified to `ValidatorConfig`.

File: lib/src/config/config.rs [hunk 3]
```diff
@@ -560,6 +711,67 @@ impl ClientConfigBuilder {
...
+        }
+
+        // Configure custom seeds
+        for seed in &config_file.network.seed_nodes {
+            self.seed(Seed::try_from(seed.clone())
+                .map_err(|e| Error::config_error(format!("Invalid seed: {:?}: {}", seed, e)))?);
+        }
+
         // Configure validator
         if config_file.validator.is_some() {
```
Comment (paberr) [inline lib/src/config/config.rs:776]: This probably needs to go behind a feature flag as the `validator` field itself is behind a feature flag, too.

Comment (paberr) [reply lib/src/config/config.rs:776]: `ValidatorConfig {}` can be simplified to `ValidatorConfig`.

File: metrics-server/src/lib.rs [hunk 4]
```diff
@@ -59,20 +59,44 @@ pub mod server;
...
-        .serve(move || {
-            server::MetricsServer::new(
-                vec![
-                    Arc::new(CM::new(consensus.blockchain.clone())),
-                    Arc::new(MempoolMetrics::new(consensus.mempool.clone())),
-                    Arc::new(NetworkMetrics::new(consensus.network.clone()))
-                ],
-                attributes!{ "peer" => consensus.network.network_config.peer_address() },
-            password.clone())
+
```
Comment (paberr) [inline metrics-server/src/lib.rs:62]: Not related to your code, but as a reminder: Jeff should port his changes from core-rs that switch the metrics server to HTTPS.

File: rpc-server/Cargo.toml [hunk 5]
```diff
@@ -47,4 +47,8 @@ nimiq-network-primitives = { path = "../network-primitives", version = "0.1" }
 nimiq-primitives = { path = "../primitives", version = "0.1" }
 nimiq-transaction = { path = "../primitives/transaction", version = "0.1" }
 nimiq-utils = { path = "../utils", version = "0.1", features = ["merkle", "time", "otp"] }
+nimiq-validator = { path = "../validator", version = "0.1", optional = true}
 nimiq-wallet = { path = "../wallet", version = "0.1" }
+
+[features]
+validator = ["nimiq-validator"]
```
Comment (paberr) [inline rpc-server/Cargo.toml:54]: Missing newline at end of file!

File: rpc-server/src/handlers/block_production_albatross.rs [hunk 6]
```diff
@@ -2,33 +2,52 @@
...
-            "proofOfKnowledge" => hex::encode(&self.pok),
+            "publicKey" => self.validator.validator_key.public.to_string(),
+        })
+    }
+
+    fn proof_of_knowledge(&self, _params: &[JsonValue]) -> Result<JsonValue, JsonValue> {
+        let key_pair = &self.validator.validator_key;
+
+        // Compute proof of knowledge.
+        // TODO: Do we need this at all? This is only needed to sign staking transactions, and
```
Comment (paberr) [inline rpc-server/src/handlers/block_production_albatross.rs:36]: I agree with this.

File: rpc-server/src/lib.rs [hunk 7]
```diff
@@ -30,6 +32,8 @@ use json::JsonValue;
 
 use crate::error::Error;
 pub use crate::handler::Handler;
+use futures::IntoFuture;
+use crate::handlers::block_production_albatross::BlockProductionAlbatrossHandler;
```
Comment (paberr) [inline rpc-server/src/lib.rs:36]: Shouldn't this be behind a feature flag as well?

**Review by paberr — DISMISSED** [review]
There are still some minor bugs in it (especially w.r.t. the feature flags), but you can also fix them in the next PR.

**Review by paberr — APPROVED** [review]

**Pull request merged**